WeakAuras/WeakAuras.lua

```diff
@@ -1642,9 +1642,127 @@ do
   end
 end
 
+local recurseDiff
+function recurseDiff(ours, theirs, ignored)
+  for key, ourVal in pairs(ours) do
+    local ignoredVal = type(ignored) == 'table' and ignored[key]
+    if ignoredVal ~= true then
+      if type(ourVal) == 'table' then
+        local theirVal = theirs[key]
+        if type(theirs[key]) ~= 'table' 
+        or recurseDiff(ourVal, theirVal, ignoredVal or nil) then
+          return true
+        end
+      elseif theirs[key] ~= ourVal then
+        return true
+      end
+    end
+  end
+end
+
+local ignoredForDiffChecking = {
+  init_started = true,
+  yOffset = true,
+  xOffset = true,
+  anchorPoint = true,
+  id = true,
+  uid = true, -- honestly could leave it unignored but whatever
+  internalVersion = true,
+  width = true,
+  frameStrata = true,
+  anchorFrameType = true,
+  selfPoint = true,
+  height = true,
+  fixedWidth = true,
+  controlledChildren = true,
+}
+
+function WeakAuras.diff(ours, theirs, ignored)
+  if not ours or not theirs then return end
+  return recurseDiff(ours, theirs, ignoredForDiffChecking)
+end
+
+function WeakAuras.MatchData(data, children) 
+  local oldData, uids = db.displays[data.id], {}
+  if data.uid then uids[data.uid] = 0 end
+  local oldChildren
+  if children then
+    oldChildren = {}
+    for i, child in ipairs(children) do
+      if child.uid then uids[child.uid] = i end
+      oldChildren[i] = db.displays[child.id]
+    end
+  end
+  for id, existingData in pairs(db.displays) do
+    local uindex = existingData.uid and uids[existingData.uid]
+    if uindex == 0 then
+      match = true
+      oldData = existingData
+    elseif uindex then
+      match = true
+      oldChildren[uindex] = existingData
+    end
+  end
+  if match then
+    local modified = WeakAuras.diff(oldData, data)
+    if children then
+      local info = {
+        unmodified = oldData and not modified and 1 or 0,
+        modified = modified and 1 or 0,
+        added = not modified and not oldData and 1 or 0,
+        deleted = 0,
+      }
+      local oldTotal = oldData and oldData.controlledChildren and #oldData.controlledChildren or 0
+      for i, child in ipairs(children) do
+        if not oldChildren[i] then
+          modified = true
+          info.added = info.added + 1
+        else
+          if WeakAuras.diff(oldChildren[i], child) then
+            modified = true
+            info.modified = info.modified + 1
+            oldTotal = oldTotal - 1
+          else
+            child.same = true
```

Comment:
I don't like that this meta-data about the data is stored inside the data, I would have expected that in info.

Reply:
Yes, that’s a hack that I forgot to improve before opening they PR.